src/mlwm/build_inference_artifact.py

```diff
@@ -0,0 +1,350 @@
+        datastore_paths[None] = _make_abspath(
+            nl_config["datastore"]["config_path"]
+        )
+    elif "datastores" in nl_config:
+        for name, datastore_config in nl_config["datastores"].items():
+            datastore_paths[name] = _make_abspath(
+                datastore_config["config_path"]
+            )
+    else:
+        raise ValueError("No datastore found in the config file.")
+
+    return datastore_paths
+
+
+def _extract_stats(datastore_config_path: str, artifact_path: str):
+    """
+    Extract statistics from the training dataset created from the datastore
+    configuration file. The statistics are saved to a file named
+    {datastore_name}.stats.zarr in the artifact output directory.
+
+    Parameters
+    ----------
+    datastore_config_path : str
+        Path to the datastore configuration file. The name of the
+        datastore is assumed to be the name of the file without the extension, i.e.
+        {datastore_name}.yaml.
+    artifact_path : str
+        Path to the artifact output directory. The statistics will be saved
+        to a file named {datastore_name}.stats.zarr in this directory.
+    """
+    fp_stats = (
+        Path(artifact_path)
+        / Path(datastore_config_path).with_suffix(".stats.zarr").name
+    )
+
+    dataset_path = Path(datastore_config_path).with_suffix(".zarr")
+    ds = xr.open_zarr(dataset_path)
+
+    ds_stats = xr.Dataset()
+
+    # XXX: this is a massive hack, but for now we will just assume that any
+    # variable that has "__" in its name twice is a statistic. This could be
+    # done better by using the implementation in mllam-data-prep...
+
+    for var_name in ds.data_vars:
+        if len(var_name.split("__")) == 3:
+            # this is a statistic
+            ds_stats[var_name] = ds[var_name].compute()
+
+    # copy the attributes too
+    ds_stats.attrs.update(ds.attrs)
+
+    logger.info(f"Saving stats for {datastore_config_path} to {fp_stats}")
+    fp_stats.parent.mkdir(parents=True, exist_ok=True)
+    ds_stats.to_zarr(str(fp_stats), mode="w", consolidated=True)
+
+
+def _extract_stats_for_all_datastores(nl_config_path, artifact_output_path):
+    datastore_paths = _find_datastore_paths(nl_config_path)
+
+    for datastore_path in datastore_paths.values():
+        _extract_stats(datastore_path, Path(artifact_output_path) / "stats")
+
+
+def _copy_checkpoint(checkpoint_path: str, fp_checkpoint_dst: str):
+    """
+    Copy the checkpoint file to the artifact output directory.
+    Parameters
+    ----------
+    checkpoint_path : str
+        Path to the checkpoint file.
+    artifact_output_path : str
+        Path to the artifact output directory.
+    """
+    fp_checkpoint_dst = Path(fp_checkpoint_dst) / "checkpoint.pkl"
+    shutil.copyfile(checkpoint_path, fp_checkpoint_dst)
+    logger.info(f"Copied checkpoint {checkpoint_path} -> {fp_checkpoint_dst}")
+
+    # XXX: we should really validate the checkpoint here somehow...
```

Comment:
It may be dangerous to rename the checkpoint?  On the other hand, we always know what it's called.

Reply:
Yes, I thought the same. I do think it is useful always having the same name, but to not lose track of the original name I decided to create `artifact.yaml` which contains information about how the inference artifact was created (including the argv cli args that the building cli was called with, which include the original checkpoint filepath)

Reply:
In addressing your comment I realised I should expose the filenames of the artifact meta info and checkpoint filename. I did this in https://github.com/leifdenby/mlwm-deployment/commit/1b61258b658c5299c930ed38da97f26981fe17a2